scripts/run-language-tests.py

```diff
@@ -0,0 +1,82 @@
+#!/usr/bin/env python3
+import os
+import sys
+import fnmatch
+import argparse
+import subprocess
+import tempfile
+
+SCRIPT_ROOT = os.path.dirname(os.path.abspath(__file__))
+TESTS_ROOT = os.path.join(SCRIPT_ROOT, 'etch-language-tests')
+
+
+def parse_commandline():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('build_folder', type=os.path.abspath,
+                        help='The path to the build folder')
+    return parser.parse_args()
+
+
+def main():
+    args = parse_commandline()
+
+    etch_cli = os.path.join(args.build_folder, 'apps', 'etch', 'etch')
+    assert os.path.isfile(etch_cli)
+
+    tests = []
+
+    # discover all the tests
+    for root, _, files in os.walk(TESTS_ROOT):
+        for path in fnmatch.filter(files, '*.etch'):
+            full_path = os.path.join(root, path)
+            name, _ = os.path.splitext(os.path.relpath(full_path, TESTS_ROOT))
+            tests.append((name, full_path))
+
+    max_test_size = max([len(n) for n, _ in tests]) + 1
+
+    print()
+    print('=== ETCH Language Tests ===')
+    print()
+
+    # loop through and execute each one of them
+    failures = []
+    for name, path in sorted(tests, key=lambda x: x[0]):
+
+        # create a temp file and call the executable inside it
+        output_file, output_filename = tempfile.mkstemp(
```

Comment:
I'd use `TemporaryFile()` context manager to improve exception safety. https://docs.python.org/3.4/library/tempfile.html#tempfile.TemporaryFile . Then you would not need the call to `close()` and `remove()` below.

Reply:
`TemporaryFile` is racey however (and I have been bitten in the past). This obv. would be fine in this case but I had planned for this code to be parallelised in the future so might be more useful.

Reply:
The docs in the link state that it also uses mkstemp under the hood. Why is doing it this way less racey?